src/config.hpp.in

```diff
@@ -25,6 +25,7 @@
 #cmakedefine ENABLE_MKLCPU_BACKEND
 #cmakedefine ENABLE_MKLGPU_BACKEND
 #cmakedefine ENABLE_NETLIB_BACKEND
+#cmakedefine ENABLE_HALF_ROUTINES
```

Comment:
I think new flag needs to be applied to all backends, but I see changes only in cublas and mklcpu, is there a reason why it's not propagated to other backends?

Reply:
This seems to be an oversight on my part, I only tested with these two backends.

Reply:
This probable means more changes needs to be done on other backends as well. 

However, I have my reservations about disabling half precision since it is part of SYCL Specification 1.2.1 section 5.1. Could you please remind me your argument for not supporting half precision in hipSYCL?

Reply:
> I have my reservations about disabling half precision since it is part of SYCL Specification 1.2.1 section 5.1.

Maybe more importantly, it is part of the SYCL 2020 spec, C.9.1.

Reply:
If it's necessary to disable `half` for hipSYCL, can disabling be done implicitly when using hipSYCL rather than adding an explicit functionality, instead of introducing more complexity into the build system?

Reply:
@mmeterel The SYCL specification lists half support as an optional kernel feature. It is not necessary for an implementation to support it inside kernels, but the type has to exist.

@sbalint98 Could we just use the SYCL 2020 API to test `half` support instead of hardcoding via the build system? https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features

This is something that should be done in general for proper SYCL 2020 use as half is an optional device feature, and therefore support in general should be queried before using it.

Reply:
@vrpascuzzi 

> Maybe more importantly, it is part of the SYCL 2020 spec, C.9.1.

No, appendix C is the OpenCL backend specification which only applies to SYCL implementations running on top of OpenCL and following the OpenCL backend specification. It neither applies to hipSYCL's backends (which are not OpenCL) nor to DPC++ Level Zero or DPC++ CUDA.

The more relevant section is here: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features

Reply:
Thanks for the info! Then the important part of that section is:
> In order to guarantee source code portability of SYCL applications that use optional kernel features, all SYCL implementations must be able to compile device code that uses these optional features regardless of whether the implementation supports the features on any of its devices.

So is it correct to disable entirely fp16 as is done here?

Reply:
I'm not sure. The spec is a bit unclear. The existence of the type is only mentioned in the OpenCL backend spec, but the intent indeed was to allow optional kernel features inside kernels if support is queried at runtime - this requires compile-time support.

hipSYCL has the `half` type (despite it not being functional), so it should be possible to compile the interface. I think we should investigate whether it is sufficient if just test at runtime for the `fp16` device aspect, as I had proposed above.

Reply:
I think "_...all SYCL implementations..._" is very clear.

A run-time test seems fine from my point of view, so long as it compiles without modification to the source nor does it require additional CMake variables.

And by the way, thanks for hipSYCL!

Reply:
> I think "...all SYCL implementations..." is very clear.

It's not even defined in the core spec what the `fp16` optional feature means, only in the backend specification. At the same time, backend specifications are not binding for other backends. IMO there's a gap there. The definition of the features mentioned (`fp16`, `fp64`, `atomic64`) should be defined in the core spec if they are supposed to be well-defined everywhere.

Reply:
> It's not even defined in the core spec what the fp16 optional feature means...

Can it be that it's definition is inferred by the LLVM?
https://clang.llvm.org/docs/LanguageExtensions.html#half-precision-floating-point

AMDGPU is one of the currently supported targets. Out of curiosity (for my own understanding), what is the limitation in hipSYCL to support fp16? hipSYCL is also based on LLVM.

Reply:
Perhaps it'd be useful to ask @bader or @Pennycook for clarification.

Thanks for the fruitful discussion, @illuhad! Good luck with this!

Reply:
I agree that the specification is unclear, here.  The definition of `half` is something that I think needs more discussion.

There are already a few internal Khronos issues open that I think are relevant to this issue: https://gitlab.khronos.org/sycl/Specification/-/issues/322, https://gitlab.khronos.org/sycl/Specification/-/issues/455.  @illuhad, if you agree that those issues are relevant, would you mind adding some more details there?